File: android/src/main/java/com/lebrislo/bluetooth/mesh/utils/NodesOnlineStateManager.kt
package com.lebrislo.bluetooth.mesh.utils

import android.util.Log
import com.getcapacitor.JSObject
import com.lebrislo.bluetooth.mesh.BluetoothMeshPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import org.json.JSONArray
import java.util.concurrent.ConcurrentHashMap

class NodesOnlineStateManager private constructor(
    private val offlineTimeout: Long = 10_000L, // Timeout in milliseconds
) {
    private val tag: String = NodesOnlineStateManager::class.java.simpleName

    private val nodes = ConcurrentHashMap<Int, Boolean>() // Tracks node states (online/offline)
    private val heartbeatTimestamps = ConcurrentHashMap<Int, Long>() // Tracks last heartbeat timestamps
    private var monitoringJob: Job? = null

    companion object {
        @Volatile
        private var instance: NodesOnlineStateManager? = null

        fun getInstance() =
            instance ?: synchronized(this) {
                instance ?: NodesOnlineStateManager().also { instance = it }
            }
    }

    // Add a new node to the network
    fun addNode(unicastAddress: Int) {
        if (!nodes.containsKey(unicastAddress)) {
            Log.i(tag, "add node to online state manager: $unicastAddress")
            nodes[unicastAddress] = false // Initialize as offline
            heartbeatTimestamps[unicastAddress] = 0L
        }
    }

    // Remove a node from the network
    fun removeNode(unicastAddress: Int) {
        Log.i(tag, "remove node from online state manager: $unicastAddress")
        nodes.remove(unicastAddress)
        heartbeatTimestamps.remove(unicastAddress)
    }

    fun clearNodes() {
        Log.i(tag, "clear nodes from online state manager")
        nodes.clear()
        heartbeatTimestamps.clear()
    }

    fun resetStatus() {
        nodes.keys.forEach { unicastAddress ->
            nodes[unicastAddress] = false
        }
    }

    // Record a heartbeat for a node
    fun heartbeatReceived(unicastAddress: Int) {
        if (nodes.containsKey(unicastAddress)) {
            heartbeatTimestamps[unicastAddress] = System.currentTimeMillis()
        } else {
            println("Unknown node: $unicastAddress")
        }
    }

    // Start monitoring node states
    fun startMonitoring() {
        monitoringJob = CoroutineScope(Dispatchers.IO).launch {
            while (isActive) {
                val currentTime = System.currentTimeMillis()
                var hasStateChanged = false

                nodes.keys.forEach { unicastAddress ->
                    val lastHeartbeat = heartbeatTimestamps[unicastAddress] ?: 0L
                    val isOnline = currentTime - lastHeartbeat < offlineTimeout

                    if (nodes[unicastAddress] != isOnline) {
                        nodes[unicastAddress] = isOnline
                        hasStateChanged = true
                    }
                }

                // Notify the network if any state changed
                if (hasStateChanged) {
                    notifyNetwork() // Send a snapshot of the current node states
                }

                delay(1000) // Check every second
            }
        }
    }

    // Stop monitoring node states
    fun stopMonitoring() {
        monitoringJob?.cancel()
    }

    fun getNodesOnlineStates(): JSObject {
        val result = JSObject()
        val states = JSONArray()
        nodes.forEach { nodeState ->
            val state = JSObject()
            state.put("unicastAddress", nodeState.key)
            state.put("isOnline", nodeState.value)
            states.put(state)
        }
        result.put("states", states)

        return result
    }

    fun notifyNetwork() {
        val notification = getNodesOnlineStates()
        NotificationManager.getInstance()
            .sendNotification(BluetoothMeshPlugin.MESH_NODE_ONLINE_STATE_EVENT, notification)
    }
}

File: ios/Sources/BluetoothMeshPlugin/utils/NodesOnlineStateManager.swift
//
//  File.swift
//  BluetoothMeshPlugin
//
//  Created by LE BRIS Loris on 25/03/2026.
//

import Capacitor
import Foundation
import NordicMesh

final class NodesOnlineStateManager {

    static let shared = NodesOnlineStateManager()

    static let meshNodeOnlineStateEvent = "meshNodeOnlineStateEvent"

    /// Timeout offline en millisecondes (par défaut: 8000ms)
    private let offlineTimeoutMs: Int64

    /// Queue série: garantit la cohérence des dictionnaires (équivalent de ConcurrentHashMap + discipline d'accès)
    private let stateQueue = DispatchQueue(label: "com.lebrislo.bluetooth.mesh.NodesOnlineStateManager.state")

    /// Etat online/offline par nœud
    private var nodes: [UInt16: Bool] = [:]

    /// Dernier heartbeat reçu (epoch ms) par nœud
    private var heartbeatTimestampsMs: [UInt16: Int64] = [:]

    /// Timer de monitoring
    private var monitoringTimer: DispatchSourceTimer?

    private init(offlineTimeoutMs: Int64 = 8_000) {
        self.offlineTimeoutMs = offlineTimeoutMs
    }

    // MARK: - Gestion des nœuds

    func addNode(unicastAddress: UInt16) {
        stateQueue.sync {
            guard nodes[unicastAddress] == nil else { return }
            NSLog("NodesOnlineStateManager: add node \(unicastAddress)")
            nodes[unicastAddress] = false
            heartbeatTimestampsMs[unicastAddress] = 0
        }
    }

    func removeNode(unicastAddress: UInt16) {
        stateQueue.sync {
            NSLog("NodesOnlineStateManager: remove node \(unicastAddress)")
            nodes.removeValue(forKey: unicastAddress)
            heartbeatTimestampsMs.removeValue(forKey: unicastAddress)
        }
    }

    func clearNodes() {
        stateQueue.sync {
            NSLog("NodesOnlineStateManager: clear nodes")
            nodes.removeAll()
            heartbeatTimestampsMs.removeAll()
        }
    }

    /// Re-cadre tous les nœuds à offline (ne touche pas les timestamps).
    func resetStatus() {
        stateQueue.sync {
            // Snapshot des clés pour éviter tout piège d’itération + mutation
            let keys = Array(nodes.keys)
            for addr in keys {
                nodes[addr] = false
            }
        }
    }

    // MARK: - Heartbeat

    /// Appelé quand un heartbeat (ou un message servant de heartbeat) est reçu.
    func heartbeatReceived(unicastAddress: UInt16) {
        let now = Self.nowMs()
        stateQueue.sync {
            guard nodes[unicastAddress] != nil else {
                NSLog("NodesOnlineStateManager: unknown node \(unicastAddress)")
                return
            }
            heartbeatTimestampsMs[unicastAddress] = now
        }
    }

    // MARK: - Monitoring périodique

    func startMonitoring() {
        stateQueue.sync {
            guard monitoringTimer == nil else { return }

            // Le timer exécute son handler sur stateQueue => accès aux dictionnaires sans locks additionnels
            let timer = DispatchSource.makeTimerSource(queue: stateQueue)
            timer.schedule(deadline: .now(), repeating: .seconds(1))
            timer.setEventHandler { [weak self] in
                self?.tickLocked()
            }
            timer.resume()

            monitoringTimer = timer
        }
    }

    func stopMonitoring() {
        stateQueue.sync {
            monitoringTimer?.cancel()
            monitoringTimer = nil
        }
    }

    // MARK: - Snapshot JSON pour JS

    func getNodesOnlineStates() -> JSObject {
        stateQueue.sync {
            makeSnapshotLocked()
        }
    }

    /// Equivalent Kotlin notifyNetwork(): envoie un snapshot courant.
    func notifyNetwork() {
        let snapshot = getNodesOnlineStates()
        NotificationManager.shared.sendNotification(
            event: NodesOnlineStateManager.meshNodeOnlineStateEvent,
            data: snapshot
        )
    }

    // MARK: - Internals (toujours appelés sur stateQueue)

    private func tickLocked() {
        let now = Self.nowMs()
        let addresses = Array(nodes.keys)

        var hasStateChanged = false

        for addr in addresses {
            let lastHeartbeat = heartbeatTimestampsMs[addr] ?? 0
            let isOnline = (now - lastHeartbeat) < offlineTimeoutMs

            if nodes[addr] != isOnline {
                nodes[addr] = isOnline
                hasStateChanged = true
            }
        }

        if hasStateChanged {
            let snapshot = makeSnapshotLocked()
            NotificationManager.shared.sendNotification(
                event: NodesOnlineStateManager.meshNodeOnlineStateEvent,
                data: snapshot
            )
        }
    }

    private func makeSnapshotLocked() -> JSObject {
        var states: JSArray = []
        states.reserveCapacity(nodes.count)

        for (unicastAddress, isOnline) in nodes {
            let state: JSObject = [
                "unicastAddress": Int(unicastAddress),
                "isOnline": isOnline,
            ]
            states.append(state)
        }

        return ["states": states]
    }

    private static func nowMs() -> Int64 {
        Int64(Date().timeIntervalSince1970 * 1000)
    }
}

extension NodesOnlineStateManager: MeshHeartbeatDelegate {
    public func meshNetworkManager(
        _ manager: MeshNetworkManager,
        didReceiveHeartbeat heartbeat: MeshHeartbeat
    ) {

        let src = UInt16(heartbeat.source)  // Address est typiquement UInt16-like
        NodesOnlineStateManager.shared.heartbeatReceived(unicastAddress: src)
    }
}
